contracts/FreeMint.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MIT
+// Compatible with OpenZeppelin Contracts ^5.0.0
+pragma solidity ^0.8.22;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {ERC721Pausable} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721Pausable.sol";
+import {ERC721URIStorage} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/// @title FreeMint NFT Contract
+/// @notice Contract for minting NFTs with free mint capability
+contract FreeMint is ERC721, ERC721URIStorage, ERC721Pausable, Ownable {
+    using Strings for uint256;
+
+    uint256 private _nextTokenId;
+
+    string public baseContractURI;
+
+    string public baseURI;
+
+    string public baseExtension;
+
+    uint256 public maxSupply;
+
+    uint256 public mintLimit;
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _baseURI,
+        string memory _baseExtension,
+        uint256 _maxSupply,
+        uint256 _mintLimit,
+        uint256 _startWithTokenId
+    ) ERC721(_name, _symbol) Ownable(msg.sender) {
+        setBaseURI(_baseURI);
+        baseExtension = _baseExtension;
+        maxSupply = _maxSupply;
+        mintLimit = _mintLimit;
+        require(_startWithTokenId <= 1, "Invalid startWithTokenId");
+        _nextTokenId = _startWithTokenId;
+    }
+
+    function mint(address to, uint256 amount) public whenNotPaused {
+        require(amount > 0, "Invalid amount");
+        require(_nextTokenId + amount < maxSupply, "Max supply reached");
+        // Allow unlimited minting if mintLimit is 0
+        if (mintLimit != 0) {
+            require(balanceOf(to) + amount <= mintLimit, "Exceeds mint limit");
+        }
+        for (uint256 i = 0; i < amount; i++) {
+            _safeMint(to, _nextTokenId++);
+        }
+    }
+
+    //@notice bulk mint function for owner
+    function bulkMint(
+        address[] calldata recipients,
+        uint256[] calldata amounts
+    ) public onlyOwner whenNotPaused {
+        require(
+            recipients.length > 0 && amounts.length > 0,
+            "Empty arrays not allowed"
+        );
+        require(
+            recipients.length == amounts.length,
+            "Recipients and amounts length mismatch"
+        );
+
+        uint256 totalAmount = 0;
+        for (uint256 i = 0; i < amounts.length; i++) {
+            require(amounts[i] > 0, "Invalid amount");
+            totalAmount += amounts[i];
+        }
+
+        require(_nextTokenId + totalAmount <= maxSupply, "Max supply reached");
+
+        for (uint256 i = 0; i < recipients.length; i++) {
```

Comment:
I guess no need to do one loop for verification and one loop for action. You can combine both, and save gas costs:
```solidity
uint256 amountToBeMinted = 0;
for (uint256 i = 0; i < amounts.length; i++) {
   require(amounts[i] > 0, "Invalid amount");
   amountToBeMinted += amounts[i];
   require(_nextTokenId + amountToBeMinted <= maxSupply, "Max supply reached");
   
   for (uint256 j = 0; j < amounts[i]; j++) {
       _safeMint(recipients[i], _nextTokenId++);
   }
}
```

Reply:
true, it is leftover from previous version of contract. Will fix